### PR TITLE
Add support for GitHub Enterprise

### PR DESCRIPTION
**Pull request by mdwheele**
This introduces support for releasing packages on GitHub Enterprise. I have tested this at NC State and the package functions as reported through a new configuration option, `github_hostname`. There are currently no integration tests that I could find that test collaboration between `Api\GitHub` and `Http`. I can add a few to demonstrate that based on configuration, HTTP base URL is adjusted appropriately. Let me know if this should be done.

A new configuration option, `github_hostname`, is added to allow GitHub Enterprise users to specify their appliance hostname. The `Github` `AbstractApi` is responsible for taking that hostname and appending `/api/v3` if the hostname does not contain `github.com`. This is documented at https://developer.github.com/v3/enterprise. This feature is **forward compatible** (double-check please) for current GitHub.com users. If unspecified, a default of `api.github.com` is populated in config rather than `null`. **_This_** is what feels awkward to me, but I think it's a matter of opinion since at runtime the API chosen is based on project `origin` and it is completely reasonable for someone to have a user config that uses many git hosting providers, I imagine. I leave whether "this is awkward" or not to the maintainers!

A `private` method was created to set repository name on `Api\Github` based on origin. This is called in constructor. The implementation of this was changed to remove (what I saw) as superfluous code. I implemented a test for `Api\Github` that seems to correctly identify the repository name on "dot com" and Enterprise. If there are edge-cases that have been regressed by this change, I'll revert removals I've made. `git blame` on the class did not reveal any special circumstances / reasons for which this logic was chosen. Looks to just be part of the first "minimalist implementation".

Anywho, I look forward to review.

Implementation is open for adjustment, obviously.

---

Here's some sample output from our install at NC State University; pulling issues for some arbitrary repository:

```
[vagrant@local restful-misery]$ ../producer.producer/bin/producer issues
mdwheele/restful-misery

    7. GUI Workflow for Adding Shortcodes
        https://github.ncsu.edu/mdwheele/restful-misery/issues/7

    13. Using __invoke, or __construct for the wordpress classes instead of a doTheNeedful() type approach
        https://github.ncsu.edu/mdwheele/restful-misery/issues/13

    15. Auth
        https://github.ncsu.edu/mdwheele/restful-misery/issues/15

    16. Additional p and br tags added before table
        https://github.ncsu.edu/mdwheele/restful-misery/issues/16
```

### CODE REVIEW

**Comment by pmjones** [comment]
Dude, sweet! Review forthcoming.

File: src/Api/Github.php [hunk 1]
```diff
@@ -26,30 +26,32 @@ class Github extends AbstractApi
...
-        if (substr($origin, 0, $len) == $ssh) {
-            $repoName = substr($origin, $len);
-        }
-
-        // strip .git from the end
-        if (substr($repoName, -4) == '.git') {
-            $repoName = substr($repoName, 0, -4);
-        }
+        $repoName = explode(':', $repoName)[1];
+        $repoName = rtrim($repoName, '.git');
```
Comment (pmjones) [inline src/Api/Github.php]: With [`rtrim()`](http://php.net/rtrim), the second param is a character mask, not a literal string, so this is not quite the same thing as removing a literal `.git` from the end. It probably _works_ but it doesn't convey the right intent and might have unexpected side-effects.

Comment (mdwheele) [inline src/Api/Github.php]: Would following suffice?

``` php
$repoName = preg_replace('/\.git$/', '', $repoName);
```

vs current:

``` php
// strip .git from the end
if (substr($repoName, -4) == '.git') {
    $repoName = substr($repoName, 0, -4);
}
```

If not, will go ahead and drop previous implementation back into place!

**Comment by pmjones** [comment]
Very clean. The use of github_hostname is a nice bit.

I have a concern about the picking-of-the-API-object in ProducerContainer based on github_hostname. I imagine a future case where Producer will support self-installed GitLab as well, which means there may well be an equivalent gitlab_hostname. In that case, defaulting to GitHub Enterprise won't work.

Perhaps the thing to do is see if `(github|gitlab)_hostname` is set, and use that as an indication this is a self-hosted remote origin. (I am not married to the idea, just a suggestion.)

Thoughts?

File: src/Api/Github.php (hunk 1 above)
Comment (pmjones) [inline src/Api/Github.php]: No strong feelings; it's 6 of one and a half-dozen of the other.

Comment (pmjones) [inline src/Api/Github.php]: Or as my wife sometimes says, it's 6-and-a-half dozen one way.

**Comment by mdwheele** [comment]
That sounds reasonable to me, though I had another thought. Should Producer support cases where a developer has repositories that may each use different self-hosted GitHub/GitLab?

For example, I might have repository `A` that points to a GitHub Enterprise installation at `ghe.foo.com` and a repository `B` that points to a GHE installation at `ghe.bar.com`. In this case, a single `github_hostname` would not suffice.

To support this expansion, I can't think of a way to detect what you're dealing with other than polling. If not supporting this, we can stick to some sort of text-matching strategy; but I'm thinking the behaviour of having a default populated in `Config` will have to go away so that we can confidently use the presence of that assignment as an indicator of remote hosting.

GitHub Enterprise returns an X-Header (`X-GitHub-Request-Id`) with every request, so the remote `origin` could be polled and if `X-GitHub-Request-Id` is there, we return `Api\Github`. If we poll, then there's probably no need for setting a hostname in configuration other than an optimization for folks that only work with one instance.

All that said... I don't know anybody that works with more than one GHE... GitLab might be a different story considering it's open-source? 

What do you think? Go forward with supporting a single remote-hosted origin or implement some strategy for detection that isn't purely pattern-matching based?

I also suppose users could put this detail in their project configuration (if I understand how config is loaded, correctly) But yeah, I agree that how the appropriate implementation of `AbstractApi` is chosen is a little awkward.

**Comment by mdwheele** [comment]
Yeah, totally looks like the "most supportable" means would be to specify `github_hostname` in the repository Producer config, if needed. Which would put us back at pattern-matching! I'll come up with something that isn't so sore to look at. Sorry for so much "thinking out loud" above. I'm having to deal with a new Great Dane puppy while working and it's just a tad distracting haha.

**Comment by shochdoerfer** [comment]
@mdwheele I would love to find a solution that allows a per-project configuration. What about "parsing" the git repo url? Not sure if this would work for GitHub Enterprise but for GitLab this seems to a way to go. In case a http or https url is used the guess is easy. For an SSH url we would at least know the hostname of the endpoint and thus need 2 requests to figure out if https is used or http as a fallback.

**Comment by shochdoerfer** [comment]
Just realized that in GitLab you are able to define a different ssh hostname. That means my initial idea does not work for all setups. 

Why not forcing the user to add an entry in .git/config which would expose the endpoint url that we need to talk to? I mean for http or https urls we can detect the endpoint, in case of an SSH url the user would need to add the entry (or we provide a command to do so) in the .git/config file.

**Comment by mdwheele** [comment]
@shochdoerfer I believe this method does, in fact, allow per-project configuration. Let me know where I am wrong.  I am able to specify per-project remote hosting on different Enterprise installs; our production as well as a deployment appliance. Parsing will not work for Enterprise as appliances are able to be deployed to anywhere. I'm also able to specify github.com as a remote host per project. 

@pmjones, my latest commit has working code. I'd like to add some more tests, but let me know if this is a preferred direction.

**Comment by pmjones** [comment]
@mdwheele :

> Sorry for so much "thinking out loud" above.

Not at all! Creative work is often like that.

I like this direction. It seems like `isGithubBased()` could be reduced somewhat, but no big deal. Adding tests, if you can, would be good.

Thanks for this good work!

**Comment by mdwheele** [comment]
Okie doke! I've shortened the `isGethubBased()` implementation logically, now that there are tests that describe intended behaviour. I only wanted to test that `ProducerContainer` returns the correct implementation of `AbstractApi` given some different scenarios. I chose to use reflection to "unlock" the `newApi` method and call it directly given an origin and mocked `Config` object. If preferred, this logic might be extracted into a factory that implements the decision on which implementation to use. This would remove the need for reflection to open-up the protected method. I personally think this test is readable enough, but it depends on your feelings on testing state (non-visible) rather than public behaviour.

Thoughts? If looks good, I'll proceed with implementing hosted Gitlab integration i accordance with #16. I do need more feedback from @shochdoerfer on which portions of this work aren't working according to expectation. My local branch is functioning with cloud-hosted Gitlab as well as a copy running locally on my laptop, but I may be missing some details from above.

**Comment by pmjones** [comment]
> If looks good, I'll proceed with implementing hosted Gitlab integration

Proceed! @shochdoerfer advise as you can. Thanks guys!

**Comment by shochdoerfer** [comment]
@mdwheele have not checked the code. Judging from your comments I had the impression that there`s still some work to do and wanted to give advice. I am quite busy today and tomorrow, will check the code as soon as I can and provide feedback.

**Comment by mdwheele** [comment]
@shochdoerfer Sure thing. What is currently there is functional for GitHub Enterprise; so no more work to be done there. I'll be merging my patch for Gitlab onto this PR later this afternoon for review. Thanks!

**Comment by mdwheele** [comment]
Alrighty, I've added more tests that verify that for GitHub and Gitlab-based projects, the repository name can be derived from `http`, `https` and `ssh` remote names. However, the client for APIs for both is still hard-coded to use HTTPS. In my opinion, this should be forced as we're passing credentials.

`GithubTest` and `GitlabTest` contain data providers that test all combinations of remote types I could think of. You'll see `origin`, the configured hostname and the expected repository name. I'm going to leave this until @shochdoerfer can check in again. I have also tested (by hand) `https` and `ssh` addresses for GHE and GL (cloud and self-hosted). Sooooo, I think we're good! 💁

**Comment by pmjones** [comment]
I think hard-coding https is The Right Thing To Do (tm). @shochdoerfer over to you!

**Comment by shochdoerfer** [comment]
Looks good to me so far. I would propose one minor change: Instead of guessing what the GitHub / GitLab url might be why not storing the url instead of the hostname? That seems the most flexible solution for me.

**Comment by mdwheele** [comment]
Could you describe that flexibility? I'm down for whatever makes most sense.

From my own experience, here's what I have:
1. Given a hostname, I don't see much guessing as to what the API endpoint is for GitHub Enterprise. I would imagine the same applies to a stock hosted-Gitlab instance. In both cases, the endpoints are documented and can be derived from hostname alone.
2. I cannot tell you how many times I've had to remind students here that the API endpoint (and things like pages) for GitHub Enterprise is NOT the same as on GitHub.com. That said, it is easy to know which hostname your instance is on and the "rules" for deriving endpoints are well documented.
3. If left as `*_hostname` and the API endpoints change, for whatever reason, we can patch for that. We can't patch (as easily) for user / project configuration.

I'm sure there are counter-examples for `*_hostname`. The only thing that comes to mind would be things like wonky reverse proxy setups in front of the appliances such that custom endpoints are established; custom routing happens. GHE basically says, "Don't screw with the black box, folks." Gitlab is probably more "flexible" in that regard and might justify use of a `github_baseurl` or `gitlab_baseurl`. 

Also FWIW, [KnpLabs GitHub Client](https://github.com/KnpLabs/php-github-api/blob/master/lib/Github/Client.php#L80) uses `base_url` as the configuration variable for where the GitHub instance is which would be in opposition here.

Implementation-wise, theres no problem. The only change I would make is to `parse_url` to grab the `host` of `*_baseurl` and continue on as currently written.

Do you have any pros/cons you could add? I'm a bit on the fence trading off user ease in configuration with flexibility in system administration of appliances (which, in GHE's case, is not advised).

**Comment by shochdoerfer** [comment]
I realized that it is possible to install GitLab with an [relative](https://gitlab.com/gitlab-org/omnibus-gitlab/blob/master/doc/settings/configuration.md#configuring-the-external-url-for-gitlab) url. Not sure how many people do that. This might lead to errors in the repo name detection. That is why I thought it could make sense to make both settings configurable. But that might be an option we could implement later when there is the need for it.

**Comment by mdwheele** [comment]
Yeah, that might be a possibility. Kinda hard without evidence. Most folks (I imagine) are running Gitlab as a separate appliance on its own host, but that's my imagination. I've only ever installed on CentOS 7 and it kinda wants domain over the box to the point it was fighting with Apache for port 80, etc etc.

@pmjones Could you make a call on which would be preferred given all of the above and I'll make necessary changes to implement, if necessary.

**Comment by pmjones** [comment]
Command decision: skip the "custom path / relative URL" for now. I'm happy to just have self-hosted Gitlab in the first place.

**Comment by mdwheele** [comment]
With a final review, I think it's ready to go. If @shochdoerfer has time to test the work himself from the perspective of a user of Producer, that'd be awesome. I've "poked it with a stick" as best I can, but am honestly just picking up the tool for the first time.

I plan to use it a great bit for work here. Release management has always been a pain.

**Comment by pmjones** [comment]
@shochdoerfer over to you!

**Comment by shochdoerfer** [comment]
Ran a quick test against our own GitLab instance. Seems to work. @pmjones up to you to merge. Thanks @mdwheele for your work! Now we can finally start to use producer :)

**Comment by pmjones** [comment]
Thanks @mdwheele for your good work here!
